### PR TITLE
VRF-330: enabling batch fulfillment for VRF e2e tests

### PR DESCRIPTION
**Pull request by iljapavlovs**

### CODE REVIEW

File: integration-tests/contracts/ethereum_vrf_common.go [hunk 1]
```diff
@@ -106,6 +107,29 @@ func parseRequestRandomnessLogs(coordinator Coordinator, logs []*types.Log) (*Co
...
+				}
+				randomWordsFulfilledEventArr = append(randomWordsFulfilledEventArr, randomWordsFulfilledEvent)
+			}
+			if topic.Cmp(vrf_coordinator_v2.VRFCoordinatorV2RandomWordsFulfilled{}.Topic()) == 0 {
+				randomWordsFulfilledEvent, err := coordinator.ParseRandomWordsFulfilled(*eventLog)
+				if err != nil {
+					return nil, fmt.Errorf("parse RandomWordsFulfilled log failed, err: %w", err)
+				}
+				randomWordsFulfilledEventArr = append(randomWordsFulfilledEventArr, randomWordsFulfilledEvent)
+			}
```
Comment (ibrajer) [inline integration-tests/contracts/ethereum_vrf_common.go]: That looks like the same code to me. Perpahs just checking if the topic comes from V2 or V25 and if that's not the case, throw an error? So something like this:

```
if topic.Cmp(vrf_coordinator_v2_5.VRFCoordinatorV25RandomWordsFulfilled{}.Topic()) != 0 && topic.Cmp(vrf_coordinator_v2.VRFCoordinatorV2RandomWordsFulfilled{}.Topic()) != 0 {
	// return an error
}
// topic is detected, the processing is the same for both coordinator versions
randomWordsFulfilledEvent, err := coordinator.ParseRandomWordsFulfilled(*eventLog)
if err != nil {
	return nil, fmt.Errorf("parse RandomWordsFulfilled log failed, err: %w", err)
}
randomWordsFulfilledEventArr = append(randomWordsFulfilledEventArr, randomWordsFulfilledEvent)
```

**Review by ibrajer — DISMISSED** [review]

File: integration-tests/contracts/ethereum_vrf_common.go (hunk 1 above)
Comment (iljapavlovs) [reply integration-tests/contracts/ethereum_vrf_common.go]: good catch, thanks!

**Review by ibrajer — DISMISSED** [review]

**Review by ibrajer — APPROVED** [review]

File: integration-tests/smoke/vrfv2_test.go [hunk 2]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+	env, vrfContracts, vrfKey, nodeTypeToNodeMap, err = vrfv2.SetupVRFV2Universe(testcontext.Get(t), t, config, chainID, cleanupFn, newEnvConfig, l)
+	require.NoError(t, err, "Error setting up VRFv2 universe")
+
+	evmClient, err := env.GetEVMClient(chainID)
+	require.NoError(t, err, "Getting EVM client shouldn't fail")
+	defaultWalletAddress = evmClient.GetDefaultWallet().Address()
+
+	//batchMaxGas := config.MaxGasLimit() (2.5 mill) + 400_000 = 2.9 mill
+	//callback gas limit set by consumer = 500k
+	// so 4 requests should be fulfilled inside 1 tx since 500k*4 < 2.9 mill
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1244]: remove?

File: integration-tests/smoke/vrfv2_test.go [hunk 3]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+
+	t.Run("Batch Fulfillment Enabled", func(t *testing.T) {
+		configCopy := config.MustCopy().(tc.TestConfig)
+
+		vrfNode, exists := nodeTypeToNodeMap[vrfcommon.VRF]
+		require.True(t, exists, "VRF Node does not exist")
+
+		//ensure that no job present on the node
+		err = actions.DeleteJobs([]*client.ChainlinkClient{vrfNode.CLNode.API})
+		require.NoError(t, err)
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1260]: add error message?

File: integration-tests/smoke/vrfv2_test.go [hunk 4]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+		wgAllRequestsFulfilled.Wait()
+
+		l.Info().
+			Interface("Request Count", requestCount).
+			Interface("Fulfilment Count", fulfilmentCount).
+			Msg("Request/Fulfilment Stats")
+
+		clNodeTxs, resp, err := nodeTypeToNodeMap[vrfcommon.VRF].CLNode.API.ReadTransactions()
+		require.NoError(t, err)
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1339]: add error message?

File: integration-tests/smoke/vrfv2_test.go [hunk 5]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+		require.Equal(t, 200, resp.StatusCode)
+		var batchFulfillmentTxs []client.TransactionData
+		for _, tx := range clNodeTxs.Data {
+			if common.HexToAddress(tx.Attributes.To).Cmp(common.HexToAddress(vrfContracts.BatchCoordinatorV2.Address())) == 0 {
+				batchFulfillmentTxs = append(batchFulfillmentTxs, tx)
+			}
+		}
+		// verify that all fulfillments should be inside one tx
+		require.Equal(t, 1, len(batchFulfillmentTxs))
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1348]: add error message?

File: integration-tests/smoke/vrfv2_test.go [hunk 6]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+		require.Equal(t, vrfContracts.BatchCoordinatorV2.Address(), fulfillmentTXToAddress, "Fulfillment Tx To Address should be the BatchCoordinatorV2 Address when batch fulfillment is enabled")
+
+		fulfillmentTxReceipt, err := evmClient.GetTxReceipt(fulfillmentTx.Hash())
+		require.NoError(t, err)
+
+		randomWordsFulfilledLogs, err := contracts.ParseRandomWordsFulfilledLogs(vrfContracts.CoordinatorV2, fulfillmentTxReceipt.Logs)
+		require.NoError(t, err)
+
+		// verify that all fulfillments should be inside one tx
+		require.Equal(t, int(randRequestCount), len(randomWordsFulfilledLogs))
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1369]: add error message to the three above?

File: integration-tests/smoke/vrfv2_test.go [hunk 7]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+	// so 4 requests should be fulfilled inside 1 tx since 500k*4 < 2.9 mill
+
+	batchFulfilmentMaxGas := *config.VRFv2.General.MaxGasLimitCoordinatorConfig + 400_000
+	config.VRFv2.General.CallbackGasLimit = ptr.Ptr(uint32(500_000))
+
+	expectedNumberOfFulfillmentsInsideOneBatchFulfillment := (batchFulfilmentMaxGas / *config.VRFv2.General.CallbackGasLimit) - 1
+	randRequestCount := expectedNumberOfFulfillmentsInsideOneBatchFulfillment
+
+	t.Run("Batch Fulfillment Enabled", func(t *testing.T) {
+		configCopy := config.MustCopy().(tc.TestConfig)
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1253]: we don't want to run them in parallel?

File: integration-tests/smoke/vrfv2_test.go [hunk 8]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+		require.Equal(t, int(randRequestCount), len(randomWordsFulfilledLogs))
+	})
+	t.Run("Batch Fulfillment Disabled", func(t *testing.T) {
+		configCopy := config.MustCopy().(tc.TestConfig)
+
+		vrfNode, exists := nodeTypeToNodeMap[vrfcommon.VRF]
+		require.True(t, exists, "VRF Node does not exist")
+		//ensure that no job present on the node
+		err = actions.DeleteJobs([]*client.ChainlinkClient{vrfNode.CLNode.API})
+		require.NoError(t, err)
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1378]: add error message?

File: integration-tests/smoke/vrfv2_test.go [hunk 9]
```diff
@@ -1193,3 +1187,297 @@ func TestVRFV2NodeReorg(t *testing.T) {
...
+		require.Equal(t, 200, resp.StatusCode)
+
+		var singleFulfillmentTxs []client.TransactionData
+		for _, tx := range clNodeTxs.Data {
+			if common.HexToAddress(tx.Attributes.To).Cmp(common.HexToAddress(vrfContracts.CoordinatorV2.Address())) == 0 {
+				singleFulfillmentTxs = append(singleFulfillmentTxs, tx)
+			}
+		}
+		// verify that all fulfillments should be in separate txs
+		require.Equal(t, int(randRequestCount), len(singleFulfillmentTxs))
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2_test.go:1480]: add error message to requires above? 😅

File: integration-tests/smoke/vrfv2plus_test.go [hunk 10]
```diff
@@ -2083,3 +2080,287 @@ func TestVRFv2PlusNodeReorg(t *testing.T) {
...
+				//cancel subs and return funds to sub owner
+				vrfv2plus.CancelSubsAndReturnFunds(testcontext.Get(t), vrfContracts, defaultWalletAddress, subIDsForCancellingAfterTest, l)
+			}
+		}
+		if !*vrfv2PlusConfig.General.UseExistingEnv {
+			if err := env.Cleanup(test_env.CleanupOpts{TestName: t.Name()}); err != nil {
+				l.Error().Err(err).Msg("Error cleaning up test environment")
+			}
+		}
+	}
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2plus_test.go:2120]: can't we define that as a function? you could have higher order function that returns this function as well, if needed. I see it's identical to the previous tests, why copy & paste?

File: integration-tests/smoke/vrfv2plus_test.go [hunk 11]
```diff
@@ -2083,3 +2080,287 @@ func TestVRFv2PlusNodeReorg(t *testing.T) {
...
+	env, vrfContracts, vrfKey, nodeTypeToNodeMap, err = vrfv2plus.SetupVRFV2PlusUniverse(testcontext.Get(t), t, config, chainID, cleanupFn, newEnvConfig, l)
+	require.NoError(t, err, "Error setting up VRFv2Plus universe")
+
+	evmClient, err := env.GetEVMClient(chainID)
+	require.NoError(t, err, "Getting EVM client shouldn't fail")
+	defaultWalletAddress = evmClient.GetDefaultWallet().Address()
+
+	//batchMaxGas := config.MaxGasLimit() (2.5 mill) + 400_000 = 2.9 mill
+	//callback gas limit set by consumer = 500k
+	// so 4 requests should be fulfilled inside 1 tx since 500k*4 < 2.9 mill
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2plus_test.go:2137]: remove?

File: integration-tests/smoke/vrfv2_test.go (hunk 8 above)
Comment (iljapavlovs) [reply integration-tests/smoke/vrfv2_test.go:1378]: reluctantly adding those, but I dont think those add much value

File: integration-tests/smoke/vrfv2plus_test.go [hunk 12]
```diff
@@ -2083,3 +2080,287 @@ func TestVRFv2PlusNodeReorg(t *testing.T) {
 	})
 
 }
+
+func TestVRFv2PlusBatchFulfillmentEnabledDisabled(t *testing.T) {
```
Comment (Tofel) [inline integration-tests/smoke/vrfv2plus_test.go:2084]: well, tbh I would extract common logic to one function and then just pass as arguments vrfv2 and vrfv2plus specific parts so we don't have almost identical test

**Review by Tofel — APPROVED** [review]
conditional ;-)

**Pull request merged**